app/src/main/java/org/p2p/wallet/newsend/smartselection/SmartSelectionCoordinator.kt

```diff
@@ -0,0 +1,142 @@
+import kotlinx.coroutines.launch
+import org.p2p.core.token.Token
+import org.p2p.wallet.feerelayer.model.FeeRelayerFee
+import org.p2p.wallet.infrastructure.dispatchers.CoroutineDispatchers
+import org.p2p.wallet.newsend.model.FeePayerState
+import org.p2p.wallet.newsend.model.smartselection.FeePayerFailureReason
+import org.p2p.wallet.newsend.model.smartselection.SmartSelectionState
+import org.p2p.wallet.newsend.smartselection.handler.TriggerHandler
+import org.p2p.wallet.newsend.smartselection.strategy.StrategyExecutor
+
+private const val TAG = "SmartSelectionCoordinator"
+
+class SmartSelectionCoordinator(
+    private val strategyExecutor: StrategyExecutor,
+    private val triggerHandlers: List<TriggerHandler>,
+    dispatchers: CoroutineDispatchers
+) : CoroutineScope {
+
+    override val coroutineContext: CoroutineContext = SupervisorJob() + dispatchers.io +
+        Executors.newSingleThreadExecutor().asCoroutineDispatcher()
+
+    private val feePayerState = MutableStateFlow<FeePayerState>(FeePayerState.Idle)
+
+    private var recentTrigger: SmartSelectionTrigger? = null
+
+    private lateinit var feePayerToken: Token.Active
+
+    private val internalSmartSelectionState = MutableStateFlow<SmartSelectionState>(SmartSelectionState.Cancelled)
+
+    private var triggerJob: Job? = null
+
+    init {
+        launch {
+            internalSmartSelectionState
+                .stateIn(this)
+                .collectLatest { handleSmartSelectionState(it) }
+        }
+    }
+
+    fun getFeePayerStateFlow(): StateFlow<FeePayerState> = feePayerState.asStateFlow()
+
+    fun isFreeAndInputEmpty(): Boolean {
+        return feePayerState.value.isTransactionFree() && feePayerState.value.isEmptyInput()
+    }
+
+    fun getFeeData(): Pair<Token.Active, FeeRelayerFee>? {
+        return when (val currentState = feePayerState.value) {
+            is FeePayerState.CalculationSuccess -> currentState.feePayerToken to currentState.fee
+            is FeePayerState.ReduceAmount -> currentState.feePayerToken to currentState.fee
+            else -> null
+        }
+    }
+
+    fun updateFeePayer(newFeePayerToken: Token.Active) {
+        this.feePayerToken = newFeePayerToken
+    }
+
+    @Throws(IllegalStateException::class)
+    fun requireFeePayer(): Token.Active {
+        if (!::feePayerToken.isInitialized) {
+            throw IllegalStateException("Fee Payer is not initialized!")
+        }
+
+        return feePayerToken
+    }
+
+    fun onNewTrigger(newTrigger: SmartSelectionTrigger) {
+        if (recentTrigger is SmartSelectionTrigger.FeePayerManuallyChanged) {
+            // just calculate fees with current fee payer token without smart selection
+            return
+        }
+
+        recentTrigger = newTrigger
+
+        triggerJob?.cancel()
+
+        launch {
+            try {
+                triggerHandlers.forEach { it.handleTrigger(internalSmartSelectionState, newTrigger, feePayerToken) }
```

Comment:
I know we have the same approach in swap, but I still don't understand, why we allow to some handlers mutate internal state of coordinator? Why not just return value from handler and emit it to state internally? Also it would make it easier to test.

Reply:
that's a good question. 
Some handlers has several states. 
Like loading, then actiona handling, then stop loading. 
And returning the value cannot give us the expected behavior. 

Therefore I just didn't spend time for inventing smth new, and just reused. 
But I'd really glad if you'll suggest some alternatives, I don't really like the current approach as well